zip-0311.rst

```diff
@@ -0,0 +1,111 @@
+::
+
+  ZIP: 0311 
+  Title: FlyClient ZCash SPV
+  Champions: 
+  Status: Draft
+  Category: Consensus
+  Created: 2019-03-30
+  License: MIT 
+
+
+Terminology
+===========
+The key words "MUST", "SHOULD", and "MAY" in this document are to be interpreted as described in RFC 2119. [#RFC2119]_
+
+Light client
+  A client that is not a full participant in the network of Zcash peers. It can send and receive payments, but does not store or validate a copy of the blockchain.
+
+High probability
+  An event occurs with high probability if it occurs with probability 1-O(1/2^λ), where λ is the security parameter.
+
+Negligible probability
+  An event occurs with negligible probability if it occurs with probability O(1/2^λ), where λ is the security parameter.
+
+Merkle mountain range (MMR)
+  A Merkle mountain range (MMR) is binary hash tree that allows for efficient appends of new leaves without changing the value of existing nodes [#PeterTodd]_.
+
+Hard fork 
+  A hard fork is a change to a protocol that makes transactions from older clients invalid. 
+  
+
+Abstract
+========
+A proposed implementation of the probabilistic verification FlyClient protocol [#FlyClient]_ on top of the existing ZCash reference light client protocol [#ZIPXXX]_. This can be implemented with a hard fork to include a MMR proof in each block header.
+
+Motivation
+==========
+FlyClient reduces the number of block headers needed for light client verification of a valid chain, from linear (as in the current reference protocol) to logarithmic in blockchain length. This verification is correct with high probability. It also allows creation of subchain proofs, so light clients need only check blocks later than the most recently verified block index. Following that, verification of a transaction inclusion within that block follows the usual reference protocol [#ZIPXXX]_. 
+
+A smaller proof size could facilitate ‘side-chaining’ with other blockchains such as Ethereum. It also reduces bandwidth and storage requirements for limited clients like mobile or IoT devices.
```

Comment:
Without a new precompile it's not economical to evaluate a single equihash on Ethereum, which means that adding MMR commitments to zcash will not immediately enable relays. However, this is still a clear win for light clients

If we're targeting relays for the future, we should consider Ethereum's capabilities when instantiating the commitment, verifying, and sampling schemes. This likely means using `sha2` or `keccak256` to construct the MMRs (as we know both Eth1 and Eth2 will support those functions)

Reply:
It looks like a BLAKE2b precompile might get into Ethereum Istanbul: https://gitcoin.co/issue/ethereum-cat-herders/PM/64/2955

Reply:
It looks like a BLAKE2b precompile might get into Ethereum Istanbul: https://gitcoin.co/issue/ethereum-cat-herders/PM/64/2955